feat: Initial PureCN implementation

dodslaser opened the pull request:
### This PR:

Added: Rule for quantifying purity and ploidy with PureCN
Added: Rule for generating segmentation using PureCN
Added: Rule for exporting cnvkit segmentation to .seg for compatibility with PureCN
Changed: The `compile_output_list` was modified to better accommodate outputs with different structures/suffixes

I have not included any integration tests for this PR as I have not found a suitable small dataset that generates valid purity/ploidy solutions. When PureCN cannot produce a valid solution it does not output any files, which crashes the workflow. I have yet to find any real-world dataset that doesn't generate a valid solution, but for small test datasets this does happen.

Furthermore, the `compile_output_list` was modified quite heavily since adding the various outputs from PureCN to the current function would have required some pretty ugly code. If need be I can revert this and shoehorn the outputs into the current function.

### Review and tests: 
- [ ] Tests pass
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Code review
- [ ] `CHANGELOG.md` is updated
- [ ] New code is executed and covered by tests, and test approve

jonca79 commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,15 @@ cnvkit_scatter:
   container: "docker://hydragenetics/cnvkit:0.9.9"
 
+cnvkit_seg:
+  container: "docker://marrip/cnvkit:0.9.9"
```
```suggestion
  container: "docker://hydragenetics/cnvkit:0.9.9"
```

dodslaser replied:
Fixed in latest commit

jonca79 commented on workflow/rules/cnvkit.smk:
```diff
@@ -189,3 +189,38 @@ rule cnvkit_vcf:
+    params:
+        extra=config.get("cnvkit_seg", {}).get("extra", ""),
+    log:
+        "cnv_sv/cnvkit_seg/{sample}_{type}.log",
```
```suggestion
        "cnv_sv/cnvkit_seg/{sample}_{type}.seg.log",
```

jonca79 commented on workflow/rules/cnvkit.smk:
```diff
@@ -189,3 +189,38 @@ rule cnvkit_vcf:
+        "cnv_sv/cnvkit_seg/{sample}_{type}.log",
+    benchmark:
+        repeat(
+            "cnv_sv/cnvkit_seg/{sample}_{type}.benchmark.tsv",
```
```suggestion
            "cnv_sv/cnvkit_seg/{sample}_{type}.seg.benchmark.tsv",
```

jonca79 commented on workflow/rules/cnvkit.smk:
```diff
@@ -189,3 +189,38 @@ rule cnvkit_vcf:
+            config.get("cnvkit_seg", {}).get("benchmark_repeats", 1),
+        )
+    threads: config.get("cnvkit_seg", {}).get("threads", config["default_resources"]["threads"])
+    resources:
+        threads=config.get("cnvkit_seg", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("cnvkit_seg", {}).get("time", config["default_resources"]["time"]),
+        mem_mb=config.get("cnvkit_seg", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
+        mem_per_cpu=config.get("cnvkit_seg", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
+        partition=config.get("cnvkit_seg", {}).get("partition", config["default_resources"]["partition"]),
```
```suggestion
    resources:
        mem_mb=config.get("cnvkit_seg", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
        mem_per_cpu=config.get("cnvkit_seg", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
        partition=config.get("cnvkit_seg", {}).get("partition", config["default_resources"]["partition"]),
        threads=config.get("cnvkit_seg", {}).get("threads", config["default_resources"]["threads"]),
        time=config.get("cnvkit_seg", {}).get("time", config["default_resources"]["time"]),
```

jonca79 commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,15 @@ cnvkit_scatter:
+
+purecn:
+  container: "docker://markusriester/purecn:2.2.0"
+
```
add empty interval file so that the dry run goes through

jonca79 commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+            config.get("purecn_coverage", {}).get("benchmark_repeats", 1),
+        )
+    threads: config.get("purecn_coverage", {}).get("threads", config["default_resources"]["threads"])
+    resources:
+        threads=config.get("purecn_coverage", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("purecn_coverage", {}).get("time", config["default_resources"]["time"]),
+        mem_mb=config.get("purecn_coverage", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
+        mem_per_cpu=config.get("purecn_coverage", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
+        partition=config.get("purecn_coverage", {}).get("partition", config["default_resources"]["partition"]),
```
```suggestion
    resources:
        mem_mb=config.get("purecn_coverage", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
        mem_per_cpu=config.get("purecn_coverage", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
        partition=config.get("purecn_coverage", {}).get("partition", config["default_resources"]["partition"]),
        threads=config.get("purecn_coverage", {}).get("threads", config["default_resources"]["threads"]),
        time=config.get("purecn_coverage", {}).get("time", config["default_resources"]["time"]),
```

jonca79 commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+            config.get("purecn", {}).get("benchmark_repeats", 1),
+        )
+    threads: config.get("purecn", {}).get("threads", config["default_resources"]["threads"])
+    resources:
+        threads=config.get("purecn", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("purecn", {}).get("time", config["default_resources"]["time"]),
+        mem_mb=config.get("purecn", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
+        mem_per_cpu=config.get("purecn", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
+        partition=config.get("purecn", {}).get("partition", config["default_resources"]["partition"]),
```
```suggestion
    resources:
        mem_mb=config.get("purecn", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
        mem_per_cpu=config.get("purecn", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
        partition=config.get("purecn", {}).get("partition", config["default_resources"]["partition"]),
        threads=config.get("purecn", {}).get("threads", config["default_resources"]["threads"]),
        time=config.get("purecn", {}).get("time", config["default_resources"]["time"]),
```

jonca79 commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,15 @@ cnvkit_scatter:
+
+purecn_coverage:
+  container: "docker://markusriester/purecn:2.2.0"
+
```
Add required options in config and dummy files in .tests/integration
          - genome
          - segmentation_method
          - interval_padding

jonca79 commented:
For the compile_output_list, do you really need all the output files or is it enough with one main file that is always created?

jonca79 commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+rule purecn:
+    input:
+        unpack(get_purecn_inputs),
+        vcf="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz",
+        tbi="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz.tbi",
```
```suggestion       
        vcf="annotation/background_annotation/{sample}_{type}.background_annotation.include.exon.filter.snv_soft_filter.vcf.gz",
        tbi="annotation/background_annotation/{sample}_{type}.background_annotation.include.exon.filter.snv_soft_filter.vcf.gz.tbi",
```

jonca79 commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+rule purecn:
+    input:
+        unpack(get_purecn_inputs),
+        vcf="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz",
+        tbi="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz.tbi",
```
```suggestion
        vcf="snv_indels/bcbio_variation_recall_ensemble/{sample}_{type}.ensembled.vcf.gz",
        tbi="snv_indels/bcbio_variation_recall_ensemble/{sample}_{type}.ensembled.vcf.gz.tbi",
```

dodslaser replied:
This currently will not work as PureCN crashes when the way BQ is reported is not consistent between callers (lima1/PureCN#249)

jonca79 commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+rule purecn:
+    input:
+        unpack(get_purecn_inputs),
+        vcf="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz",
+        tbi="snv_indels/gatk_mutect2/{sample}_{type}.merged.softfiltered.vcf.gz.tbi",
```
Add dummy files in .tests

dodslaser replied:
Added in latest commits

Smeds commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+    container:
+        config.get("purecn_coverage", {}).get("container", config["default_container"])
+    conda:
+        "../envs/purecn_coverage.yaml"
```
```suggestion
        "../envs/purecn.yaml"
```

we should try to use one cond env file for the same tool, and not one new for each rule

Smeds commented on workflow/schemas/config.schema.yaml:
```diff
@@ -344,6 +358,103 @@ properties:
+    properties:
+      benchmark_repeats:
+        type: integer
+        description: set number of times benchmark should be repeated for purecn_coverage
```
```suggestion
        description: set number of times benchmark should be repeated
```

Smeds commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,112 @@
+        "--interval-padding={params.interval_padding} "
+        "--sampleid={wildcards.sample} "
+        "--out=cnv_sv/purecn/{wildcards.sample}_{wildcards.type} "
+        "--force --seed=1337 "
```
Is it possible that one would like to be able to change "--seed"? If that's the case we should maybe add a param for it with a default value.

Smeds replied:
If it's no point of allowing a user to modify it you can leave it as it is.

dodslaser replied:
OK, will leave it for now

Smeds commented:
> > For the compile_output_list, do you really need all the output files or is it enough with one main file that is always created?
> 
> Yes and no. The number of outputs can definitely be reduced to the .csv for the `purecn` rule and the coverage_loess.txt.gz for `purecn_coverage` rule (fixed in latest commits). However, the `purecn` rule will not produce any output file if no valid purity/ploidy solutions are found, so output is not guaranteed.

Not being guaranteed a output file will be a problem for snakemake. Would it be possible to update the shell part to work like [manta wrapper](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/manta.html), which is to:
1 - write output to a temporary directory 
2 - copy the requested output out from the temporary directory

and to handle the case with no output, when the command has run successfully, could be to create a empty file

Smeds commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,20 @@ cnvkit_scatter:
+  interval_padding: 100
+
+purecn_coverage:
+  container: "docker://markusriester/purecn:2.2.0"
```
would it be possible to build a container and upload it to hydra-genetics organization? ? 

can be done at: https://github.com/hydra-genetics/docker

dodslaser replied:
Yes, although the [Dockerfile for markusriester/purecn](https://github.com/lima1/PureCN/blob/master/Dockerfile) is a bit complicated to build support for GenomicsDB. So in that case it would probably just be `FROM markusriester/purecn:2.2.0` without much more added. Is that OK?

Smeds replied:
That's ok for me! At least as a start, in the future we could look at having a complete build file (should make it easier to update version)

dodslaser replied:
I created a PR for this (hydra-genetics/docker#95)

dodslaser commented:
> Not being guaranteed a output file will be a problem for snakemake. Would it be possible to update the shell part to work like [manta wrapper](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/manta.html), which is to: 1 - write output to a temporary directory 2 - copy the requested output out from the temporary directory
> 
> and to handle the case with no output, when the command has run successfully, could be to create a empty file

Yes this is probably the correct way of doing it.

Smeds commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,20 @@ cnvkit_scatter:
+
+purecn_coverage:
+  container: "docker://markusriester/purecn:2.2.0"
+  intervals: "/reference/targets_intervals_25000.txt"
```
is this file missing? I can't find it in the pull-request and should it be without the starting slash?

```suggestion
  intervals: "reference/targets_intervals_25000.txt"
```

Smeds replied:
is this file missing? I can't find it in the pull-request and should it be without the starting slash?

```suggestion
  intervals: "reference/targets_intervals_25000.txt"
```

Smeds commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,20 @@ cnvkit_scatter:
+
+purecn_coverage:
+  container: "docker://markusriester/purecn:2.2.0"
+  intervals: "/reference/targets_intervals_25000.txt"
```
is this file missing? I can't find it in the pull-request and should it be without the starting slash?

```suggestion
  intervals: "reference/targets_intervals_25000.txt"
```

dodslaser replied:
Good catch, that should point to a dummy file.

Smeds commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,20 @@ cnvkit_scatter:
+  interval_padding: 100
+
+purecn_coverage:
+  container: "docker://markusriester/purecn:2.2.0"
```
```suggestion
  container: "docker://hydragenetics/purecn:2.2.0"
```

Smeds commented on .tests/integration/config.yaml:
```diff
@@ -26,6 +26,20 @@ cnvkit_scatter:
+  container: "docker://hydragenetics/cnvkit:0.9.9"
+
+purecn:
+  container: "docker://markusriester/purecn:2.2.0"
```
```suggestion
  container: "docker://hydragenetics/purecn:2.2.0"
```

dodslaser commented:
Latest commits changed a few things, namely:

- Rebased against latest develop
- Reverted my changes to `compile_output_list` to fit with the existing style
- Changed `purecn` shell-directive to always produce a .csv file (will be empty if PureCN returns non-zero) 
- Put all PureCN in a temp directory and created a separate `purecn_copy_output` rule

The `purecn_copy_output` rule will wait for the main `purecn` rule (checkpoint) to complete and copy the requested files from the temporary directory to `cnv_sv/purecn/{sample}_{type}{suffix}` where possible suffixes are listed as wildcard_constraints. The file `cnv_sv/purecn/{sample}_{type}.csv` will always be generated, but will be empty if PureCN is not successful.

Smeds commented on workflow/rules/purecn.smk:
```diff
@@ -0,0 +1,129 @@
+    message:
+        "{rule}: Copy {input} to {output}"
+    shell:
+        "ln {input.file} {output} || cp {input.file} {output}"
```
why both ln and cp, is it to support platforms that doesn't have ln?

dodslaser replied:
Either that or if output is not on same device (since hard-link doesn't work across devices).

I don't think it's actually necessary here, it's just how I usually do it out of habit.
